src/prototype.rs

```diff
@@ -1546,51 +1594,223 @@ fn make_code_span(tree: &mut Tree<Item>, s: &str, open: usize, close: usize) {
+        return None;
+    }
+    let mut url = String::new();
+    while let Some(c) = scanner.next_char() {
+        match c {
+            '>' => return Some(url),
+            '\x00'..='\x1f' | '<' => return None,
+            '\\' => {
+                let c = scanner.next_char()?;
+                if !(c <= '\x7f' && is_ascii_punctuation(c as u8)) {
+                    url.push('\\');
+                }
+                url.push(c);
+            }
+            _ => url.push(c),
+        }
+    }
+    None
+}
+
+fn scan_link_destination(scanner: &mut InlineScanner) -> Option<String> {
+    let save = scanner.clone();
+    if let Some(url) = scan_link_destination_pointy(scanner) {
+        return Some(url);
+    }
+    *scanner = save;
+    scan_link_destination_plain(scanner)
+}
+
+fn scan_link_title(scanner: &mut InlineScanner) -> Option<String> {
+    let open = scanner.next_char()?;
+    if !(open == '\'' || open == '\"' || open == '(') {
+        return None;
+    }
+    let mut title = String::new();
+    let mut nest = 0;
+    while let Some(c) = scanner.next_char() {
+        if c == open {
+            if open == '(' {
+                nest += 1;
+            } else {
+                return Some(title);
+            }
+        }
+        if open == '(' && c == ')' {
+            if nest == 0 {
+                return Some(title);
+            } else {
+                nest -= 1;
+            }
+        }
+        if c == '\\' {
+            let c = scanner.next_char()?;
+            if !(c <= '\x7f' && is_ascii_punctuation(c as u8)) {
+                title.push('\\');
+            }
+            title.push(c);
+        } else {
+            title.push(c);
+        }
+    }
+    None
+}
+
+fn scan_inline_link(scanner: &mut InlineScanner) -> Option<(String, String)> {
+    if !scanner.scan_ch(b'(') {
+        return None;
+    }
+    scanner.scan_while(is_ascii_whitespace);
+    let url = scan_link_destination(scanner)?;
+    let mut title = String::new();
+    let save = scanner.clone();
+    if scanner.scan_while(is_ascii_whitespace) > 0 {
+        if let Some(t) = scan_link_title(scanner) {
+            title = t;
+            scanner.scan_while(is_ascii_whitespace);
+        } else {
+            *scanner = save;
+        }
+    }
+    if !scanner.scan_ch(b')') {
+        return None;
+    }
+    Some((url, title))
+}
+
+// TODO: if it's just node, get rid of struct. But I think there probably will be
+// state related to normalization.
+struct LinkStackEl {
+    node: usize,
+}
+
+/// Handle inline HTML, code spans, and links.
 ///
 /// This function handles both inline HTML and code spans, because they have
-/// the same precedence. (Maybe rename?)
-fn handle_inline_html(tree: &mut Tree<Item>, s: &str) {
+/// the same precedence. It also handles links, even though they have lower
+/// precedence, because the URL of links must not be processed.
+fn handle_inline_pass1(tree: &mut Tree<Item>, s: &str) {
```

Comment:
I feel like this functions and many other functions in this module really dig deep into the internals of `Tree`. Is there not a way to abstract this out to the tree module?

Reply:
I'd like to do that, yes. Part of the reason this hasn't happened yet is that it hasn't been 100% clear exactly what access to the tree is needed.